lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -142,7 +143,10 @@ namespace {
       // Apply the standard SIL loop canonicalization transformations.  This
       // automatically gives us the following invariants: loops are guaranteed
       // to have a single preheader, a single backedge block, and exit??
-      canonicalizeAllLoops(&DI, &LI);
+      if (canonicalizeAllLoops(&DI, &LI)) {
+        // Recalculate PDI if canonicalization made any changes.
+        PDI.recalculate(*F);
```

Comment:
we already have `PDI.recalculate(*F);` when `if (loopChanged)`. why do we need this additional call?

Reply:
This is to take care of changes introduced by canonicalizeAllLoops, which happens before our transformations. There is no support to incrementall update PDI unlike DI and LI.

Reply:
It seems we are not using `PDI` before this call. should we remove `PDI(F)` from the init list in c'tor, so that the call here becomes its initialization? That would make the code more readable for me, but you can decide.

also consider moving the `canonicalizeAllLoops()` call to the c'tor (so that `PDI` initialization code is done as part of initialization).

Reply:
I also noticed that processAcyclicRegion() and processAcyclicRegionExcludingEnd() might not need to be public.

Reply:
i also noticed processAcyclicRegionExcludingEnd() may not need to be public.

Reply:
I will leave the PDI as is for now. I plan to move the SingleExitTransformer class to a separate file as it is getting larger. I will do some of the refactoring then.

I made the non-interface methods private.